configuration/configurator/ConfigSchemaHelper.cpp

```diff
@@ -0,0 +1,898 @@
+    CElement *pElement = NULL;
+
+    if (pAttribute == NULL)
+    {
+        pElement = m_pSchemaMapManager->getElementFromXPath(pXPath);
+        assert(pElement != NULL);
+        return pElement->getEnvValueFromXML();
+    }
+    else
+    {
+        assert(pAttribute != NULL);
+        if (nRow == 1)
+            return pAttribute->getEnvValueFromXML();
+        else
+        {
+            StringBuffer strXPath(pXPath);
+            StringBuffer strXPathOrignal(pXPath);
+
+            CConfigSchemaHelper::stripXPathIndex(strXPath);
+
+            strXPath.appendf("[%d]", nRow);
+            char pTemp[64];
+            int offset = strXPath.length() - (strlen(itoa(nRow, pTemp, 10)) - 1);
+
+            strXPath.append(strXPathOrignal, strXPath.length(), strXPathOrignal.length()-offset);
+
+            pAttribute = m_pSchemaMapManager->getAttributeFromXPath(strXPath.str());
+
+            if (pAttribute == NULL)
+                return NULL;
+
+            return pAttribute->getEnvValueFromXML();
+        }
+    }
+}
+
+int CConfigSchemaHelper::getElementArraySize(const char *pXPath) const
+{
+    assert(pXPath != NULL);
+    assert(m_pSchemaMapManager != NULL);
+
+    //CElementArray *pElementArray = m_pSchemaMapManager->getElementArrayFromXSDXPath(pXPath);
+    CElementArray *pElementArray = m_pSchemaMapManager->getElementArrayFromXPath(pXPath);
+
+    if (pElementArray == NULL)
+        return 0;
+
+    //return pElementArray->getCountOfSiblingElements(pXPath);
+    VStringBuffer strXPath("%s[1]",pElementArray->getXSDXPath());
+    //return pElementArray->getCountOfSiblingElements(pElementArray->getXSDXPath());
+    return pElementArray->getCountOfSiblingElements(strXPath.str());
+}
+
+const char* CConfigSchemaHelper::getAttributeXSDXPathFromEnvXPath(const char* pEnvXPath) const
+{
+    assert(pEnvXPath != NULL && *pEnvXPath != 0);
+    assert(m_pSchemaMapManager != NULL);
+    CAttribute *pAttribute = m_pSchemaMapManager->getAttributeFromXPath(pEnvXPath);
+
+    assert(pAttribute != NULL);
+    return pAttribute->getXSDXPath();
+}
+
+const char* CConfigSchemaHelper::getElementArrayXSDXPathFromEnvXPath(const char* pXSDXPath) const
+{
+    assert(pXSDXPath != NULL);
+    assert(m_pSchemaMapManager != NULL);
+    CElementArray *pElementArray = m_pSchemaMapManager->getElementArrayFromXSDXPath(pXSDXPath);
+
+    assert(pElementArray != NULL);
+    return pElementArray->getXSDXPath();
+}
+
+void CConfigSchemaHelper::appendAttributeXPath(const char* pXPath)
+{
+    m_strArrayEnvXPaths.append(pXPath);
+}
+
+void CConfigSchemaHelper::appendElementXPath(const char* pXPath)
+{
+    m_strArrayEnvXPaths.append(pXPath);
+}
+
+int CConfigSchemaHelper::stripXPathIndex(StringBuffer &strXPath)
+{
+    int nLen = strXPath.length()-3;
+    int nLengthOfStringInBracket = 3;
+
+    while (nLen > 0)
+    {
+        if (strXPath[nLen] == '[')
+        {
+            strXPath.reverse().remove(0,strXPath.length()-nLen).reverse();
```

Comment:
this seems extraneous, the reverse call copies the buffer content. Can we instead iterate backwards from nLen until we find the first '[' and setLength(index -1) ?